File: packages/go/whatsapp-go/plan.go
package whatsapp

import (
	"fmt"
	"time"
)

var CenarioBaseCategories = map[MessageCategory]CategoryQuota{
	CategoryAuth:      {MonthlyQuota: 40_000, UnitCostUSD: 0.0315, FreeAllowance: 0},
	CategoryUtility:   {MonthlyQuota: 60_000, UnitCostUSD: 0.0350, FreeAllowance: 0},
	CategoryService:   {MonthlyQuota: 5_000, UnitCostUSD: 0.0300, FreeAllowance: ServiceFreeAllowance},
	CategoryMarketing: {MonthlyQuota: 10_000, UnitCostUSD: 0.0625, FreeAllowance: 0},
}

var CenarioBaseVolume = UsageCounts{
	Auth: 40_000, Utility: 60_000, Service: 5_000, Marketing: 10_000,
}

var CenarioBasePlan = PlanQuotas{
	Categories: cloneQuotas(CenarioBaseCategories),
	UsdToBrl:   DefaultUsdToBrl,
	QuotaMode:  QuotaHard,
}

func MergePlan(overrides *PlanOverride, quotaMode QuotaMode) PlanQuotas {
	cats := cloneQuotas(CenarioBaseCategories)
	usd := DefaultUsdToBrl
	mode := QuotaHard
	if overrides != nil {
		if overrides.UsdToBrl != nil {
			usd = *overrides.UsdToBrl
		}
		if overrides.QuotaMode != "" {
			mode = overrides.QuotaMode
		}
		for k, patch := range overrides.Categories {
			cur := cats[k]
			if patch.MonthlyQuota != nil {
				cur.MonthlyQuota = *patch.MonthlyQuota
			}
			if patch.UnitCostUSD != nil {
				cur.UnitCostUSD = *patch.UnitCostUSD
			}
			if patch.FreeAllowance != nil {
				cur.FreeAllowance = *patch.FreeAllowance
			}
			cats[k] = cur
		}
	}
	if quotaMode != "" {
		mode = quotaMode
	}
	return PlanQuotas{Categories: cats, UsdToBrl: usd, QuotaMode: mode}
}

func QuotaFor(plan PlanQuotas, cat MessageCategory) CategoryQuota {
	return plan.Categories[cat]
}

func BillingPeriod(at time.Time, timeZone string) string {
	if timeZone == "" {
		timeZone = "America/Sao_Paulo"
	}
	loc, err := time.LoadLocation(timeZone)
	if err != nil {
		loc = time.UTC
	}
	local := at.In(loc)
	return fmt.Sprintf("%04d-%02d", local.Year(), local.Month())
}

func VencimentoForPeriod(period string, day int) (string, error) {
	var year, month int
	if _, err := fmt.Sscanf(period, "%d-%d", &year, &month); err != nil || year == 0 || month == 0 {
		return "", fmt.Errorf("periodo invalido: %s", period)
	}
	last := time.Date(year, time.Month(month)+1, 0, 0, 0, 0, 0, time.UTC).Day()
	chosen := last
	if day > 0 {
		chosen = day
		if chosen > last {
			chosen = last
		}
	}
	return fmt.Sprintf("%s-%02d", period, chosen), nil
}

func cloneQuotas(in map[MessageCategory]CategoryQuota) map[MessageCategory]CategoryQuota {
	out := make(map[MessageCategory]CategoryQuota, len(in))
	for k, v := range in {
		out[k] = v
	}
	return out
}

File: packages/nodejs/whatsapp-sdk/src/plan.ts
import type {
  CategoryQuota,
  CategoryQuotaMap,
  DeepPartialPlan,
  MessageCategory,
  PlanQuotas,
  UsageCounts,
} from "./types.js";

/**
 * Câmbio implícito na tabela do Cenário Base (ex.: US$ 1.260 → R$ 6.930).
 * Configurável por tenant/plano — não é taxa fixa de produto.
 */
export const DEFAULT_USD_TO_BRL = 5.5;

/**
 * Quotas e custos unitários do Cenário Base (~100k transações/mês).
 *
 * | Categoria     | Volume | Unitário USD | Total USD | Total BRL |
 * |---------------|--------|--------------|-----------|-----------|
 * | auth          | 40.000 | 0.0315       | 1.260     | 6.930     |
 * | utility       | 60.000 | 0.0350       | 2.100     | 11.550    |
 * | service       | 5.000  | 0.0300       | 120 (após 1k free) | 660 |
 * | marketing     | 10.000 | 0.0625       | 625       | (câmbio do plano) |
 */
export const CENARIO_BASE_CATEGORIES: CategoryQuotaMap = {
  auth: { monthlyQuota: 40_000, unitCostUsd: 0.0315, freeAllowance: 0 },
  utility: { monthlyQuota: 60_000, unitCostUsd: 0.035, freeAllowance: 0 },
  service: { monthlyQuota: 5_000, unitCostUsd: 0.03, freeAllowance: 1_000 },
  marketing: { monthlyQuota: 10_000, unitCostUsd: 0.0625, freeAllowance: 0 },
};

export const CENARIO_BASE_VOLUME: UsageCounts = {
  auth: 40_000,
  utility: 60_000,
  service: 5_000,
  marketing: 10_000,
};

export const CENARIO_BASE_PLAN: PlanQuotas = {
  categories: CENARIO_BASE_CATEGORIES,
  usdToBrl: DEFAULT_USD_TO_BRL,
  quotaMode: "hard",
};

export const SERVICE_FREE_ALLOWANCE = 1_000;

export function emptyUsageCounts(): UsageCounts {
  return { auth: 0, utility: 0, service: 0, marketing: 0 };
}

export function mergePlan(overrides?: DeepPartialPlan, quotaMode?: PlanQuotas["quotaMode"]): PlanQuotas {
  const categories = { ...CENARIO_BASE_CATEGORIES };
  if (overrides?.categories) {
    for (const key of Object.keys(overrides.categories) as MessageCategory[]) {
      const patch = overrides.categories[key];
      if (!patch) continue;
      categories[key] = { ...categories[key], ...patch };
    }
  }
  return {
    categories,
    usdToBrl: overrides?.usdToBrl ?? DEFAULT_USD_TO_BRL,
    quotaMode: quotaMode ?? overrides?.quotaMode ?? "hard",
  };
}

export function quotaFor(plan: PlanQuotas, category: MessageCategory): CategoryQuota {
  return plan.categories[category];
}

export function billingPeriod(date: Date = new Date(), timeZone = "America/Sao_Paulo"): string {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone,
    year: "numeric",
    month: "2-digit",
  }).formatToParts(date);
  const year = parts.find((p) => p.type === "year")?.value;
  const month = parts.find((p) => p.type === "month")?.value;
  if (!year || !month) {
    return `${date.getUTCFullYear()}-${String(date.getUTCMonth() + 1).padStart(2, "0")}`;
  }
  return `${year}-${month}`;
}

export function lastDayOfPeriod(period: string): string {
  const [year, month] = period.split("-").map((v) => Number(v));
  if (!year || !month) {
    throw new Error(`periodo invalido: ${period}`);
  }
  const last = new Date(Date.UTC(year, month, 0)).getUTCDate();
  return `${period}-${String(last).padStart(2, "0")}`;
}

export function vencimentoForPeriod(period: string, day?: number): string {
  const [year, month] = period.split("-").map((v) => Number(v));
  if (!year || !month) {
    throw new Error(`periodo invalido: ${period}`);
  }
  const last = new Date(Date.UTC(year, month, 0)).getUTCDate();
  const chosen = day == null || day <= 0 ? last : Math.min(day, last);
  return `${period}-${String(chosen).padStart(2, "0")}`;
}
